submission aid

guschmue opened the pull request.

christ1ne commented:
I'd like to recommend some minor changes:
1. more consistent naming to match with the reference code:
choose exact text (upper case & lower case): SingleStream,MultiStream,Server,Offline & resnet50|mobilenet|ssd-mobilenet|ssd-resnet34|gnmt
If you agree @guschmue , I will open a PR to update the submission doc to incorporate this.

2. tf-gpu_reference.json should be renamed to tf-gpu_reference_Server.json per the submission doc. (missing scenario)

3. tf_gpu.json is not matching with the rules: https://github.com/mlperf/policies/blob/master/submission_rules.adoc#56-system_desc_idjson-metadata 
https://github.com/mlperf/policies/blob/master/submission_rules.adoc#57-system_desc_id_implementation_id_scenariojson-metadata

christ1ne commented:
4 mlperf_log_trace.json is missing in most scenarios, esp. in the performance numbers.
5. shall we have a user.conf in the example even if it is empty.

nvpohanh commented on v0.5/tools/submission/submission-checker.py:
```diff
@@ -0,0 +1,238 @@
+    args = get_args()
+
+    script_path = os.path.dirname(sys.argv[0])
+    with open(os.path.join(script_path, "meta.json"), "r") as f:
```
What is this `meta.json`? I can't find it in submission rules: https://github.com/mlperf/policies/blob/master/submission_rules.adoc

guschmue replied:
Just a json template of the meta fields defined in the submission doc. 
Christine noted above ... my meta.json is a little behind the submission doc - will fix it today.

nvpohanh replied:
That makes sense. Closing.

nvpohanh commented on v0.5/tools/submission/submission-checker.py:
```diff
@@ -0,0 +1,238 @@
+def check_accuracy(model, dir):
+    is_valid = False
+    # look for: accuracy=... or mAP=...
+    with open(os.path.join(dir, "accuracy.txt"), "r") as f:
```
What is this `accuracy.txt`? I can't find it in submission rules: https://github.com/mlperf/policies/blob/master/submission_rules.adoc

guschmue replied:
Its is the output of the accuracy-[coco|imagenet].py script. 
It is called from the sample submission notebook. 
I might change the code a little to actually create the file but than I need to have coco and imagenet datasets passed into the checker.

nvpohanh replied:
Or change `accuracy-[coco|imagenet].py` so that they store accuracy results to disk? I am worried about calling those scripts from here because each implementation might have different flags which are passed into the accuracy scripts.

nvpohanh replied:
Otherwise, we will at least have to provide users with a way to specify which flags they want to pass into accuracy scripts.

guschmue replied:
The user need to run those scripts anyway, else he would would not know if he met accuracy.
I can add a message with some instructions.

guschmue replied:
I changed to print instructions to run those scripts if I can't find accuracy.txt

nvpohanh replied:
this plan sounds good.

nvpohanh commented on v0.5/tools/submission/submission-checker.py:
```diff
@@ -0,0 +1,238 @@
+            if m:
+                is_valid = True
+                break
+                # TODO: we might as well check the total count
```
We need to match GNMT accuracy result as well: `r"BLEU: ([0-9\.]+)"`

guschmue replied:
yes

nvpohanh commented on v0.5/tools/submission/submission-checker.py:
```diff
@@ -0,0 +1,238 @@
+                break
+                # TODO: we might as well check the total count
+    # check if there are any errors in the detailed log
+    fname = os.path.join(dir, "mlperf_log_detail.txt")
```
`mlperf_log_detail.txt` should be in performance directory, not accuracy directory, according to https://github.com/mlperf/policies/blob/master/submission_rules.adoc

guschmue replied:
loadgen will put it there and I'd rather have it to make sure accuracy run had no error. We should change the submission doc.

nvpohanh replied:
Makes sense now. Closing

guschmue replied:
https://github.com/mlperf/policies/issues/6

guschmue commented:
@christ1ne 
For 1: Yeah, we should use the same terms ... SingleStream,MultiStream,Server,Offline is what the ref app uses and what is in mlperf.conf.
The names of the networks - the ref app uses the detailed names but the results directory maps them to the names in the rules doc. This is because in the open division it might not be the specific model (like ssd-mobilenet) so we use the more abstract names like ssd-small.
I will check on 2,3 and 5.
For 4: We decided no submission of the trace required.

riking commented on SubmissionExample.ipynb:
````diff
@@ -0,0 +1,592 @@
+    "\n",
+    "You run this notebook from the root of the mlperf inference tree that you cloned with\n",
+    "```\n",
+    "git clone https://github.com/mlperf/inference.git --depth 1\n",
````
after #477, let's change this to check out the release tag

riking commented on SubmissionExample.ipynb:
```diff
@@ -0,0 +1,592 @@
+   "source": [
+    "### Run benchmarks using the reference implementation\n",
+    "\n",
+    "Lets pretent we want to submit mobilenet on a desktop machine with a Gtx-1080 gpu using tensorflow. \n",
```
nit: "pretend"
nit: "GTX-1080" or "gtx-1080"

psyhtest replied:
Pretend? We are going to do it for real :)

guschmue replied:
changed the text to ```Lets prepare a submission```

riking commented on SubmissionExample.ipynb:
```diff
@@ -0,0 +1,592 @@
+    "        \"accelerator_interconnect_topology\": \"-\",\n",
+    "        \"accelerator_memory_capacity\": \"8GB\",\n",
+    "        \"accelerator_memory_configuration\": \"none\",\n",
+    "        \"accelerator_model_name\": \"Gtx-1080\",\n",
```
same capitalization nit on model name

riking commented on SubmissionExample.ipynb:
```diff
@@ -0,0 +1,592 @@
+    "## we keep a copy under /data so just symlink it.\n",
+    "mkdir data\n",
+    "ln -s  /data/imagenet2012 data/\n",
+    "ln -s  /data/coco data/"
```
this may be worth calling out or mentioning as something you probably need to change

nvpohanh commented on v0.5/tools/submission/submission-checker.py:
```diff
@@ -0,0 +1,238 @@
+                name = os.path.join(device, model, scenario)
+                acc_path = os.path.join("results", device, model, scenario, "accuracy")
+                if not files_exists(files.get(acc_path), REQUIRED_ACC_FILES):
+                    bad_submissions[name] = "{} has file missmatch".format(acc_path)
```
Also, the message `... has file mismatch` is too vague. Better message would be: `Cannot find the required file "<filename>" in <dir>`

guschmue replied:
ok, I print the difference now.

nvpohanh commented on v0.5/tools/submission/submission-checker.py:
```diff
@@ -0,0 +1,238 @@
+VALID_MODELS = ["ssd-small", "ssd-large", "mobilenet", "resnet", "gnmt"]
+REQUIRED_DIRECTORIES = ["code", "results", "measurements", "systems"]
+REQUIRED_PERF_FILES = ["mlperf_log_accuracy.json", "mlperf_log_summary.txt", "mlperf_log_detail.txt"]
+REQUIRED_ACC_FILES = REQUIRED_PERF_FILES + ["accuracy.txt"]
```
Are `"mlperf_log_accuracy.json", "mlperf_log_summary.txt", "mlperf_log_detail.txt"` required in accuracy directory? what should we do if we run the LoadGen in `SubmissionRun` mode?

guschmue replied:
I think we want them so we can check for errors so I filed:
https://github.com/mlperf/policies/issues/6

nvpohanh commented on v0.5/tools/submission/submission-checker.py:
```diff
@@ -0,0 +1,238 @@
+REQUIRED_DIRECTORIES = ["code", "results", "measurements", "systems"]
+REQUIRED_PERF_FILES = ["mlperf_log_accuracy.json", "mlperf_log_summary.txt", "mlperf_log_detail.txt"]
+REQUIRED_ACC_FILES = REQUIRED_PERF_FILES + ["accuracy.txt"]
+REQUIRED_MEASURE_FILES = ["mlperf.conf", "README.md"]
```
`mlperf.conf` should be `user.conf` according to https://github.com/mlperf/policies/blob/master/submission_rules.adoc

guschmue replied:
I think it wants both so I added a empty user.conf

nvpohanh commented on v0.5/tools/submission/submission-checker.py:
```diff
@@ -0,0 +1,238 @@
+log = logging.getLogger("main")
+
+VALID_MODELS = ["ssd-small", "ssd-large", "mobilenet", "resnet", "gnmt"]
+REQUIRED_DIRECTORIES = ["code", "results", "measurements", "systems"]
```
Is this checked anywhere?

guschmue replied:
now checking VALID_MODELS. REQUIRED_DIRECTORIES  I don't need ... implicit in the code.

nvpohanh commented:
>     1. tream,MultiStream,Server,Offline & resnet50|mobilenet|ssd-mobilenet|ssd-resnet34|gnmt

@christ1ne you mentioned `resnet50|mobilenet|ssd-mobilenet|ssd-resnet34|gnmt` here but the submission rules said `{ mobilenet, ssd-small, resnet, ssd-large, gnmt }`. Should I file an issue on the submission rules? Thanks

@DilipSequeira FYI.

christ1ne commented:
@nvpohanh @DilipSequeira please go with the existing submission doc on that: { mobilenet, ssd-small, resnet, ssd-large, gnmt }.

guschmue commented:
I think I addressed most of the review feedback. 
Working on 1 more change to checker for org name and division so that a resulting PR will fit into the submission repo.

guschmue commented:
Added support for the full repo structure.
I'm not planning to add more to this PR.

psyhtest commented:
@aaronzhongii I do get an empty `mlperf_log_accuracy.json` in performance runs too.

psyhtest commented:
Actually, I'm rather wary about only `mlperf_log_accuracy.json` required. How do we know this file was obtained from exactly the same system as the performance counterpart? What's stopping a rogue submitter submitting an accuracy run from one system and a performance run from a different system?

psyhtest commented on v0.5/tools/submission/submission-checker.py:
```diff
@@ -0,0 +1,289 @@
+    meta_errors = check_system_desc_id(good_submissions, systems_json)
+
+    # 3. check measurement and code dir
+    measurement_errros = check_measurement_dir(good_submissions, systems_imp_json)
```
errors (also elsewhere)

aaronzhongii commented:
> @aaronzhongii I do get an empty `mlperf_log_accuracy.json` in performance runs too.

they are empty

aaronzhongii commented:
> Actually, I'm rather wary about only `mlperf_log_accuracy.json` required. How do we know this file was obtained from exactly the same system as the performance counterpart? What's stopping a rogue submitter submitting an accuracy run from one system and a performance run from a different system?

there is no point to check the empty file. submission checker is not a auditing tools, its purpose is to make sure the submission files conforms to submission rules. Results can be fabricated to conform to any file checker, only auditing with onsite result reproduction is the correct way to make sure no one is cheating.

nvpohanh commented on v0.5/tools/submission/submission-checker.py:
```diff
@@ -26,6 +26,13 @@
+    "resnet50": 1024,
+    "ssd-mobilenet": 256,
+    "ssd-resnet34": 64,
+    "gnmt": 3003,
```
Should this be 3003 or 3903900? @DilipSequeira @christ1ne

guschmue replied:
Yes, https://github.com/mlperf/inference_policies/blob/master/inference_rules.adoc#41-benchmarks wants 3903900 ... is this correct ?

guschmue replied:
I changed it to 3903900 to align with the rules doc.

guschmue commented:
@aaronzhongii for the code directory we check for this:
```
code_dir = os.path.join(parts[0], parts[1], "code", model, impl)
```
aka division/submitter/code/model/imp_name

For 
```
log.warning("{} contains errors".format(fname))
```
yes I agree. It is actually fixed in lg but the commit is not required for submission.

The submission checker is not really meant as audit tool. We want to catch errors before the submission. To easy to miss things.

aaronzhongii commented:
> imp_name

def check_measurement_dir(good_submissions, systems_imp_json):
    errors = []
    for submission in good_submissions:

aaronzhongii commented:
> @aaronzhongii for the code directory we check for this:
> 
> ```
> code_dir = os.path.join(parts[0], parts[1], "code", model, impl)
> ```
> 
> aka division/submitter/code/model/imp_name
def check_measurement_dir(good_submissions, systems_imp_json):
    errors = []
    for submission in good_submissions:
... 
          for i in files:
            if i.startswith(system_desc) and i.endswith(".json"):
                system_file = i
...
        impl = system_file[len(system_desc) + 1:-5]

and as in https://github.com/mlperf/policies/blob/master/submission_rules.adoc#552-inference
measurements/
   <system_desc_id>/
      <benchmark>/
         <scenario>
            <system_desc_id>_<implementation_id>_<scenario>.json
​​
thus system_file are filled with
<system_desc_id>_<implementation_id>_Offline.json etc

impl = system_file[len(system_desc) + 1:-5]
which only abandon the <system_desc_id> part and .json in the end, the remainder is
<implementation_id>_<scenario>

psyhtest commented:
Congratulations folks, now we have 8 commits on top of the previously approved `6122045`. (This PR should have really been squashed.)

Seriously, we are treating the tree as sacred, not allowing bug fixes like #437, #479, #481, #489, and then add 8 commits? (To avoid any doubt, my preference would be to merge everything essential.)

jveejay commented:
Thanks for raising this Anton, we should address this during our meeting
today.

On Tue, Oct 8, 2019 at 4:57 AM Anton Lokhmotov <notifications@github.com>
wrote:

> Congratulations folks, now we have 8 commits on top of the previously
> approved 6122045. (This PR should have really been squashed.)
>
> Seriously, we are treating the tree as sacred, not allowing bug fixes like
> #437 <https://github.com/mlperf/inference/pull/437>, #479
> <https://github.com/mlperf/inference/pull/479>, #481
> <https://github.com/mlperf/inference/pull/481>, #489
> <https://github.com/mlperf/inference/pull/489>, and then add 8 commits?
> (To avoid any doubt, my preference would be to merge everything essential.)
>
> —
> You are receiving this because you modified the open/close state.
> Reply to this email directly, view it on GitHub
> <https://github.com/mlperf/inference/pull/473?email_source=notifications&email_token=ABT6DFHAL37ORWW2U62ZYDLQNRDQZA5CNFSM4I4EWLG2YY3PNVWWK3TUL52HS4DFVREXG43VMVBW63LNMVXHJKTDN5WW2ZLOORPWSZGOEATOLBI#issuecomment-539420037>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/ABT6DFE67NUBYIN5IQCUN33QNRDQZANCNFSM4I4EWLGQ>
> .
>


-- 
Vijay Janapa Reddi, Ph. D. | Associate Professor
33 Oxford St. | Maxwell Dworkin 147 | Cambridge, MA 02138
John A. Paulson School of Engineering and Applied Sciences | Harvard
University
Email: vj@eecs.harvard.edu | Web:
http://scholar.harvard.edu/vijay-janapa-reddi

psyhtest commented:
Thanks @jveejay.

psyhtest commented:
I'll just post some recommended changes here:
```
diff --git a/v0.5/tools/submission/submission-checker.py b/v0.5/tools/submission/submission-checker.py
index ba57446..8f22a35 100755
--- a/v0.5/tools/submission/submission-checker.py
+++ b/v0.5/tools/submission/submission-checker.py
@@ -148,7 +148,7 @@ def check_results_dir(dir, filter_submitter):
                 system_id_json = os.path.join(division, submitter, "systems", system_desc + ".json")
                 device_bad = not os.path.exists(system_id_json)
                 for model in list_dir(results_path, system_desc):
-                    if model not in VALID_MODELS:
+                    if division == "closed" and model not in VALID_MODELS:
                         bad_submissions[os.path.join(system_desc, model)] = \
                             "{} has an invalid model name {}".format(os.path.join(results_path, system_desc), model)
                         log.error("{} has an invalid model name {}".format(os.path.join(results_path, system_desc), model))
```

psyhtest commented:
Unintentional errors in the tree root (e.g. an empty directory passed to `--input`) or organization name (`--submitter`) leads to successful validation:
```
$ mkdir /tmp/empty
$ python3 submission-checker.py --input /tmp/empty --submitter d1v1d1t1
INFO:main:SUMMARY: submission looks OK
```

psyhtest commented:
An unknown model name (for the open) gives an error.
```
diff --git a/v0.5/tools/submission/submission-checker.py b/v0.5/tools/submission/submission-checker.py
index ba57446..fbcaa57 100755
--- a/v0.5/tools/submission/submission-checker.py
+++ b/v0.5/tools/submission/submission-checker.py
@@ -103,7 +103,7 @@ def check_performance_dir(model, dir):
             if m:
                 rt[m.group(1).strip()] = m.group(2).strip()
 
-    if int(rt['performance_sample_count']) < PERFORMANCE_SAMPLE_COUNT[model]:
+    if int(rt['performance_sample_count']) < PERFORMANCE_SAMPLE_COUNT.get(model, 0):
         log.error("{} performance_sample_count should be {}".format(fname, PERFORMANCE_SAMPLE_COUNT[model]))
         is_valid = False
```

psyhtest commented:
@guschmue Please see a couple of issues here.

guschmue commented:
fair, let me see if I can get those fixes in.

psyhtest commented:
Thanks @guschmue.

Perhaps the latter can be better addressed with:
```
diff --git a/v0.5/tools/submission/submission-checker.py b/v0.5/tools/submission/submission-checker.py
index 49f1fa5..9108b3a 100755
--- a/v0.5/tools/submission/submission-checker.py
+++ b/v0.5/tools/submission/submission-checker.py
@@ -103,7 +103,7 @@ def check_performance_dir(model, dir):
             if m:
                 rt[m.group(1).strip()] = m.group(2).strip()

-    if int(rt['performance_sample_count']) < PERFORMANCE_SAMPLE_COUNT[model]:
+    if model in VALID_MODELS and int(rt['performance_sample_count']) < PERFORMANCE_SAMPLE_COUNT[model]:
         log.error("{} performance_sample_count should be {}".format(fname, PERFORMANCE_SAMPLE_COUNT[model]))
         is_valid = False
```
While `division` is not available in that function, if the earlier check (`if division == "closed" and model not in VALID_MODELS`) passes, then a model in the closed will be in `VALID_MODELS`.

psyhtest commented:
Another thing: I've removed my performance experiments for testing, but the check still passed.
